haystack/document_stores/opensearch.py

```diff
@@ -798,6 +1009,136 @@ def _get_raw_similarity_score(self, score):
+                        f"code_size parameter for product quantization must be less than or equal to 8. "
+                        f"Provided code_size is `{code_size}`."
+                    )
+
+                # see FAISS doc for details: https://github.com/facebookresearch/faiss/wiki/FAQ#can-i-ignore-warning-clustering-xxx-points-to-yyy-centroids
+                min_points_per_cluster = 39
+                n_clusters = 2**code_size
+                if len(documents) < n_clusters:
+                    raise DocumentStoreError(
+                        f"PQ training requires the number of training samples to be greater than or "
+                        f"equal to the number of clusters. Number of provided training samples is `{len(documents)}` "
+                        f"and the number of clusters is `{n_clusters}`."
+                    )
+                if len(documents) < n_clusters * min_points_per_cluster:
+                    logger.warning(
+                        "Consider increasing the number of training samples to at least "
+                        "`%i` to get a reliable PQ index.",
+                        n_clusters * min_points_per_cluster,
+                    )
+
+                encoder = {"name": "pq", "parameters": {"m": m, "code_size": code_size}}
+                settings += f" m:{m} code_size:{code_size}"
+                training_req_body["method"]["parameters"]["encoder"] = encoder
+
+            training_req_body["description"] = settings
+            logger.info("Training IVF index '%s' using {len(documents)} embeddings.", index)
+            train_endpoint = f"/_plugins/_knn/models/{index}-ivf/_train"
+            response = self.client.transport.perform_request(
+                "POST", url=train_endpoint, headers=headers, body=training_req_body
+            )
+            ivf_model = response["model_id"]
+
+            # Wait until model training is finished, _knn_model_trained uses a retry decorator
+            if self._knn_model_trained(ivf_model, headers=headers):
+                logger.info("Training of IVF index '%s' finished.", index)
+
+            # Delete temporary training index
+            self.client.indices.delete(index=f".{index}_ivf_training", headers=headers)
+
+            # Clone original index to temporary one
+            self.client.indices.add_block(index=index, block="read_only")
+            self.client.indices.clone(
+                index=index, target=f".{index}_temp", body={"settings": {"index": {"blocks": {"read_only": False}}}}
+            )
+            self.client.indices.put_settings(index=index, body={"index": {"blocks": {"read_only": False}}})
+            self.client.indices.delete(index=index)
+
+            # Reindex original index to newly created IVF index
+            self._create_document_index(index_name=index, headers=headers)
+            self.client.reindex(body={"source": {"index": f".{index}_temp"}, "dest": {"index": index}})
```

Comment:
I guess this can take very long exceeding the `timeout` we set during init. I digged a bit and found:https://github.com/opensearch-project/opensearch-py/blob/58aa041db325cd9c9d273f21c00c969f272e933c/opensearchpy/transport.py#L457
So by setting `params={"request_timeout":24*60*60}` we could increase it to 24 hours, which should work in most cases.

Reply:
I added the `request_timeout` param as suggested. Still, I think this is rather an edge case that the reindexing would take so long, as the usual workflow would be to first train the index and only then add the documents.